### PR TITLE
feat(vlm): integrate packed CP and vision sharding for Qwen3.5-MoE

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
> < />!NOTE< />  <br>**Final integration PR.** NVIDIA-NeMo/Automodel#2763, NVIDIA-NeMo/Automodel#2937, NVIDIA-NeMo/Automodel#2989, and NVIDIA-NeMo/Automodel#2990 are merged. This branch includes latest `main` at `d3f3b2b6d`, uses `distributed.multimodal.vision.frame_sharding` with `mesh_dims: [cp]`, and is mergeable at final head `9b24aa970`. GitHub shows 25 integration-owned changed files.

# What does this PR do?

Integrate packed context parallelism and frame-level CP vision sharding into Qwen3.5-MoE VLM training, with Qwen3.5-122B-A10B at 128K context as the large-scale target.

The implementation follows the merged unified CP contract from NVIDIA-NeMo/Automodel#2937: `prepare_model_inputs_for_cp` is metadata/sharder-only, while embedding, trainable vision encoding, multimodal splicing, and primary sequence sharding happen inside the model forward per microbatch.

# Changelog

* Select a contiguous block-diagonal `ContextParallelSharder` for packed Qwen3.5-MoE SDPA batches; ordinary batches retain the merged round-robin CP path.
* Run full-sequence multimodal embedding inside `forward`, then shard the primary embeddings contiguously for packed CP while auxiliary tensors use the identical layout.
* Route Qwen3.5-MoE full attention through block-diagonal CP SDPA and pass typed packed-document state to the GDN/linear-attention path, composing packing + GDN + CP.
* Consume NVIDIA-NeMo/Automodel#2990's frame-level vision sharding for the Qwen3.5-MoE trainable vision tower.
* Preserve temporary CP-SDPA suspension during activation-checkpoint recompute.
* Handle Qwen3.5-122B split MTP expert checkpoint layout and rebuild loaded expert tensors on the EP mesh.
* Add typed, recipe-configurable video processing so Qwen video sampling and total `T × H × W` resize volume can be bounded explicitly.
* Add the Qwen3.5-122B-A10B EP8/CP32 128K example with trainable vision, full activation checkpointing, FP32 master weights, and FP32 optimizer moments.
* Add focused model, distributed, activation-checkpoint, state-dict, processor, and recipe-contract coverage.

# Landing status

All prerequisite PRs are merged. This branch includes latest `main` (`d3f3b2b6d`) at final head `9b24aa970`; the remaining 25-file diff is the Qwen3.5-MoE packed VLM integration described below.

# Validation

## Rebase/refactor validation

After rebuilding on the merged NVIDIA-NeMo/Automodel#2937/NVIDIA-NeMo/Automodel#2989 design:

* 129 focused block-diagonal CP, activation-checkpoint, Qwen3.5-MoE, state-dict, and 122B recipe tests passed; 51 CUDA-only/unavailable-dependency tests skipped.
* 156 unified-sharder, CP utility, and VLM recipe regression tests passed; 7 CUDA-only tests skipped.
* The final video-processor and CP32 recipe changes passed 125 targeted tests; 3 CUDA-only tests skipped.
* Ruff formatting/checks, shell syntax checks, and `git diff --check` passed.
* Final [post-NVIDIA-NeMo/Automodel#2990](<https://github.com/post-NVIDIA-NeMo/Automodel/issues/2990>) merge validation at `ef2225f2a`: 456 focused CPU tests passed and 51 CUDA-only tests skipped; Ruff formatting/checks and `git diff --check` passed. The local MDX parser validated all 452 pages; the Fern CLI is not installed locally, so CI supplies the final `fern check`. GitHub reports this PR mergeable with 25 changed files.
* Latest-main merge validation at `9b24aa970`: 172 directly affected AutoPipeline, Qwen3.5-MoE, and VLM recipe tests passed; 8 CUDA-only cases skipped. Ruff formatting/checks and `git diff --check` passed. The resolution preserves upstream mRoPE PP chunking alongside the CP/packing integration.

The previously validated two-GPU packed full-attention + GDN parity run also passed (Slurm job `14214890`).

### 100-step mixed-workload run

A 100-step end-to-end CP16 run completed successfully on the pre-rebase integration source revision `7ce8dc955`; the history rebase preserves that integration logic while adapting it to the merged sharder API:

* 16 nodes / 128 H100 GPUs
* EP8, CP16, inferred DP8
* trainable vision tower with full activation checkpointing
* frame-level CP vision sharding
* neat-packed 128K sequences
* mixed text, image, and bounded-video data, including genuine examples of approximately 120K tokens
* TE FusedAdam with FP32 master weights and FP32 first/second moments
* 100/100 steps completed; no OOM, collective hang, activation-checkpoint error, non-finite gradient, or traceback
* peak allocated memory: approximately 64.2 GiB/GPU
* final aggregate throughput: 6,340 tokens/s (49.53 tokens/s/GPU)
* loss decreased from 1.7345 to effectively zero with finite gradients

[Weights & Biases run](<https://wandb.ai/Nemo-automodel/yuhez_workspace/runs/hun4pet2>)

The validation set has 48 packed sequences and was intentionally reused for the 100-step mechanics/convergence smoke. The near-zero terminal loss demonstrates that the configuration can train and overfit this dataset; it is not evidence of production-scale convergence or generalization.

### Post-refactor CP32 runs

Two exact post-refactor implementation runs used revision `df4c51ac9` and the CP32/GBS4 settings now published by `ded4f01eb`:

| Workload | Slurm | Result | Peak allocated | Steady measured step | Evidence |
| -- | -- | -- | -- | -- | -- |
| Public MedPix, 128K packs | `14293833` | 2/2 steps, finite loss and gradients | 59.62 GiB/GPU | \~92 s | [W&B](<https://wandb.ai/Nemo-automodel/yuhez_workspace/runs/rq2j94d6>) |
| Mixed text/image/video, including an approximately 121.5K-token example | `14293834` | 2/2 steps, finite loss and gradients | 58.49 GiB/GPU | \~74 s | [W&B](<https://wandb.ai/Nemo-automodel/yuhez_workspace/runs/sgwduznm>) |

Both runs used 16 nodes / 128 H100 GPUs, EP8, CP32, inferred DP4, a trainable vision tower, full activation checkpointing, and FP32 optimizer state. Video preprocessing was bounded to at most 8 sampled frames and a total sampled volume of 524,288 pixels (approximately 8 × 256 × 256).

The first logged step in both runs took approximately five minutes because it included compilation and autotuning; it is not representative steady-state throughput. The public MedPix workload is also intentionally vision-heavy—its 128K packs combine many independent single-image samples—so its approximately 3K aggregate steady TPS is a topology/correctness result, not a throughput claim. For representative long-text or bounded-video training, use the smallest CP degree that fits; CP16 preserves DP8 on this 128-GPU allocation and was more throughput-efficient than CP32/DP4.

# Before ready for review

- [X] Rebase onto the merged NVIDIA-NeMo/Automodel#2937/NVIDIA-NeMo/Automodel#2989 CP design.
- [X] Remove obsolete merge commits and already-landed dependency history.
- [X] Resolve the unified-sharder/in-forward embedding conflicts.
- [X] Rerun focused CPU regression, formatting, and lint checks.
- [X] Confirm DCO sign-off on all integration commits.
- [X] Run post-refactor CP32 public-image and mixed-video GPU validation.
- [X] Merge NVIDIA-NeMo/Automodel#2990.
- [X] Merge latest `main` (`d3f3b2b6d`), preserve upstream Qwen3.5 mRoPE PP chunking, and confirm the remaining diff is integration-only.
- [ ] Run final GitHub CI on exact head `9b24aa970`.

# Additional information

* Tracking issue: NVIDIA-NeMo/Automodel#2985
* Prerequisites: NVIDIA-NeMo/Automodel#2763, NVIDIA-NeMo/Automodel#2937, NVIDIA-NeMo/Automodel#2989, and NVIDIA-NeMo/Automodel#2990 are merged.

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
Scoped NeMo CI revalidation passed on exact head `dd4c77769e5533d9bd863963544dc4b47a3fe515`:

- [Parent pipeline 59328557](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/59328557)
- [VLM leaf pipeline 59329560](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/59329560)
- [Exact `qwen3_5_122b_128k_ep8cp32` job 372189477](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/372189477)

The exact 16-node / 128-H100 EP8+CP32 128K recipe completed all 10 steps with Slurm `COMPLETED`. It exercised both the CP32-sharded vision tower and the compact packed-document mask path. Loss remained finite and moved from 2.2999 at step 0 to 1.5559 at step 9; peak reported memory was 60.55 GiB.

This rerun also validates the fix in `dd4c77769`: the previous exact run materialized and copied a dense 128K boolean attention mask (16 GiB/GPU), then OOMed entering step 3. The corrected run skips that mask, rebuilds block-diagonal CP attention from compact document IDs, and passed the former failure point plus all remaining steps.

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/guides/cp-vision-frame-sharding.mdx and docs/model-coverage/vlm/qwen/qwen3-5-vl.mdx. No changes needed. LGTM.

File: nemo_automodel/recipes/vlm/finetune.py [hunk 1]
```diff
@@ -646,6 +648,7 @@ def setup(self):
                     batch_size=self.cfg.get("step_scheduler.local_batch_size", 1),
                     dataset_build_context=validation_build_context,
                     get_rope_index=get_rope_index,
+                    cp_size=self.mesh_context.cp_size,
```
Comment (akoumpa) [inline nemo_automodel/recipes/vlm/finetune.py:651]: nice

**Review by akoumpa — DISMISSED** [review]

**Review by athitten — APPROVED** [review]
Thank you @yuhezhang-ai ! Just re-approving based on @akoumpa 's approval
